irr/include/coreutil.h

```diff
@@ -75,9 +76,9 @@ inline io::path &deletePathFromFilename(io::path &filename)
 
 	if (p != s) {
 		++p;
-		filename = p;
```

Comment:
I was initially stumped as to what's going on, but I'm fairly sure now that this is the culprit. Because here, we're assigning *a substring of the string's contents to the string itself*!

https://github.com/minetest/minetest/blob/bca44574d511ed49d50c4eaf8f5927063f7d40ed/irr/include/irrString.h#L167-L185

This first calculates the length of that substring. *Then it resizes.* After the resize, accessing `p`, especially characters after the new length, is essentially UaF.

So what's happening is that you have some funny path like `C:\funny path` and then it wants to assign `C:\funny path` = `funny path` essentially. So it first resizes to `C:\funny p`, then it copies, but this includes the new null terminator and you get a cut off `funny p` which doesn't match the supposed length.
I assume that on Linux, it is handling the resizing differently. Perhaps it's allocating a new buffer, copying the substring over there, and freeing the old one, then using the old one after it has been freed, so there is no issue, while on Windows it reuses the old buffer and just adds the null terminator there, hence you get what I described here. Either way this shouldn't be terribly bad; you just end up with a bit of gibberish or a cut off string.

Now, why did the Irrlicht code do that? Turns out Irrlicht would not resize eagerly before, but this was changed [in a refactor](https://github.com/minetest/irrlicht/commit/07fd32da5062ab219218b3d1cae6d79103ad2c3b#diff-c59d87ef1e2533e59f015208d02c1551667c3faed4b33460ebaeb1a2da6649ecL342).